akka-http-core/src/test/scala/akka/http/javadsl/HttpExtensionApiSpec.scala

```diff
@@ -0,0 +1,447 @@
+import scala.concurrent.duration._
+import scala.util.Try
+
+class HttpExtensionApiSpec extends WordSpec with Matchers with BeforeAndAfterAll with ScalaFutures {
+
+  // tries to cover all surface area of javadsl.Http
+
+  type Host = String
+  type Port = Int
+
+  implicit val system = {
+    val testConf = ConfigFactory.parseString("""
+    akka.loggers = ["akka.testkit.TestEventListener"]
+    akka.loglevel = ERROR
+    akka.stdout-loglevel = ERROR
+    windows-connection-abort-workaround-enabled = auto
+    akka.log-dead-letters = OFF
+    akka.http.server.request-timeout = infinite""")
+
+    ActorSystem(getClass.getSimpleName, testConf)
+  }
+  implicit val materializer = ActorMaterializer()
+
+  val http = Http.get(system)
+  val connectionContext = ConnectionContext.noEncryption()
+  val serverSettings = ServerSettings.create(system)
+  val poolSettings = ConnectionPoolSettings.create(system)
+  val loggingAdapter = NoLogging
+
+  val successResponse = HttpResponse.create().withStatus(200)
+
+  val httpSuccessFunction = new Function[HttpRequest, HttpResponse] {
+    @throws(classOf[Exception])
+    override def apply(param: HttpRequest): HttpResponse = successResponse
+  }
+
+  val asyncHttpSuccessFunction = new Function[HttpRequest, CompletionStage[HttpResponse]] {
+    @throws(classOf[Exception])
+    override def apply(param: HttpRequest): CompletionStage[HttpResponse] =
+      CompletableFuture.completedFuture(successResponse)
+  }
+
+  "The Java HTTP extension" should {
+
+    // all four bind method overloads
+    "properly bind a server 1" in {
+      val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
+      val probe = TestSubscriber.manualProbe[IncomingConnection]()
+      val binding = http.bind(host, port, materializer)
+        .toMat(Sink.fromSubscriber(probe), Keep.left)
+        .run(materializer)
+      val sub = probe.expectSubscription()
+      val address = waitFor(binding).localAddress
+      sub.cancel()
+    }
+
+    "properly bind a server 2" in {
+      val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
+      val probe = TestSubscriber.manualProbe[IncomingConnection]()
+      val binding = http.bind(host, port, connectionContext, materializer)
+        .toMat(Sink.fromSubscriber(probe), Keep.left)
+        .run(materializer)
+      val sub = probe.expectSubscription()
+      val address = waitFor(binding).localAddress
+      sub.cancel()
+    }
+
+    "properly bind a server 3" in {
+      val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
+      val probe = TestSubscriber.manualProbe[IncomingConnection]()
+      val binding = http.bind(host, port, connectionContext, serverSettings, materializer)
+        .toMat(Sink.fromSubscriber(probe), Keep.left)
+        .run(materializer)
+      val sub = probe.expectSubscription()
+      val address = waitFor(binding).localAddress
+      sub.cancel()
+    }
+
+    "properly bind a server 4" in {
```

Comment:
why are these tests replicated 4 times? Isn't it better to have one test case that loops four times (maybe calling `info()` to log progress)?